### PR TITLE
Geometry rewrite

### LINKED ISSUE #40: Converting from geometry.hpp to internal representation
Currently, [convert.cpp](https://github.com/mapbox/geojson-vt-cpp/blob/master/src/convert.cpp) code converts from GeoJSON data in RapidJSON representation to GeoJSON-VT internal representation, which is implemented using `Projected*` types (basically, OGC features with a third coordinate for simplification value + bounding boxes, and additionally storing polygon area + line length for each linear ring).

To switch to a geometry.hpp-powered GeoJSON-VT-C++ in Mapbox GL Native, we need to make it possible to create GeoJSON-VT index from structures in the code directly, without the GeoJSON step. Currently this is done in hacky way by exposing `Projected*` classes from GeoJSON-VT and then reconstructing them in https://github.com/mapbox/mapbox-gl-native/blob/master/src/mbgl/annotation/shape_annotation_impl.cpp.

Instead we would need to either have separate `convert_geojson` and `convert_geometry` code paths, or just fully ditch GeoJSON conversion and only accept geometry.hpp for the input, moving the GeoJSON parsing responsibility to GL Native.

cc @artemp @jfirebaugh

### PR DESCRIPTION
**Pull request by mourner**
A work in progress to close #40.

- [x] address @jfirebaugh's comments
- [x] readd `getTile` method
- [x] update glfw viz app
- [x] replace the build system with simple Makefile
- [x] get travis passing again
- [x] update tests
- [ ] ~~try readding breadth-first drilldown using a stack rather than recursion~~ (not important)
- [x] rebase

### CODE REVIEW

File: include/mapbox/geojsonvt.hpp [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
+#include <mapbox/geojsonvt/convert.hpp>
+#include <mapbox/geojsonvt/tile.hpp>
+#include <mapbox/geojsonvt/types.hpp>
+#include <mapbox/geojsonvt/wrap.hpp>
 
-#include <mutex>
-#include <string>
+#include <chrono>
+#include <cmath>
+#include <iostream>
```
Comment (kkaefer) [inline include/mapbox/geojsonvt.hpp]: don't include `iostream` if not strictly necessary

File: include/mapbox/geojsonvt.hpp [hunk 2]
```diff
@@ -35,40 +37,90 @@ struct Options {
...
+            const double tolerance =
+                (z == options.maxZoom ? 0 : options.tolerance / (z2 * options.extent));
+
+            tiles.emplace(
+                id, detail::Tile{ features, z, x, y, options.extent, options.buffer, tolerance });
+            stats[z] = (stats.count(z) ? stats[z] + 1 : 1);
+            total++;
+        }
+
+        auto& tile = tiles.find(id)->second;
```
Comment (kkaefer) [inline include/mapbox/geojsonvt.hpp]: This find-or-create-if-missing pattern can be rewritten more efficiently like this:

```
auto it = tiles.find(id);
if (it == tiles.end()) {
    it = tiles.emplace(id, ...).first;
}
auto& tile = it->second;
```

This avoids a second lookup in both cases.

File: include/mapbox/geojsonvt/clip.hpp [hunk 3]
```diff
@@ -1,46 +1,221 @@
...
 namespace mapbox {
 namespace geojsonvt {
+namespace detail {
+
+inline vt_line_string newSlice(vt_multi_line_string& parts, vt_line_string& slice, double dist) {
+    if (!slice.empty()) {
+        slice.dist = dist;
+        parts.push_back(std::move(slice));
+    }
+    return {};
```
Comment (kkaefer) [inline include/mapbox/geojsonvt/clip.hpp]: This always returns an empty object?

File: include/mapbox/geojsonvt/clip.hpp [hunk 4]
```diff
@@ -1,46 +1,221 @@
...
+}
+
+template <uint8_t I>
+inline void
+clipLine(const vt_line_string& line, double k1, double k2, vt_multi_line_string& slices) {
```
Comment (kkaefer) [inline include/mapbox/geojsonvt/clip.hpp]: mark all parameters as `const` which are not getting modified (`k1`, `k2`)

File: include/mapbox/geojsonvt/clip.hpp [hunk 5]
```diff
@@ -1,46 +1,221 @@
...
+            }
+            clipped.emplace_back(result, props);
+
+        } else if (geom.is<vt_multi_polygon>()) {
+            vt_multi_polygon result;
+            for (const auto& polygon : geom.get<vt_multi_polygon>()) {
+                vt_polygon p;
+                for (const auto& ring : polygon) {
+                    const auto new_ring = clipRing<I>(ring, k1, k2);
+                    if (!new_ring.empty())
```
Comment (kkaefer) [inline include/mapbox/geojsonvt/clip.hpp]: braces

File: include/mapbox/geojsonvt/clip.hpp [hunk 6]
```diff
@@ -1,46 +1,221 @@
...
+            if (parts.size() == 1)
+                clipped.emplace_back(parts[0], props);
+            else
+                clipped.emplace_back(parts, props);
+
+        } else if (geom.is<vt_polygon>()) {
+            vt_polygon result;
+            for (const auto& ring : geom.get<vt_polygon>()) {
+                const auto new_ring = clipRing<I>(ring, k1, k2);
+                if (!new_ring.empty())
```
Comment (kkaefer) [inline include/mapbox/geojsonvt/clip.hpp]: braces

File: include/mapbox/geojsonvt/convert.hpp [hunk 7]
```diff
@@ -1,72 +1,128 @@
...
+    const size_t len = ring.size();
+    result.reserve(len);
+
+    for (const auto& p : ring) {
+        result.push_back(project(p, tolerance));
+    }
+
+    double area = 0.0;
+
+    for (size_t i = 0; i < len - 1; ++i) {
```
Comment (kkaefer) [inline include/mapbox/geojsonvt/convert.hpp]: overflow if `len == 0` (also in many other places)

File: include/mapbox/geojsonvt/tile.hpp [hunk 8]
```diff
@@ -1,43 +1,189 @@
...
+          z2(std::pow(2, z_)),
+          sq_tolerance(tolerance_ * tolerance_) {
+
+        for (const auto& feature : source) {
+            const auto& geom = feature.geometry;
+            const auto& props = feature.properties;
+
+            num_points += feature.num_points;
+
+            if (geom.is<vt_point>()) {
```
Comment (kkaefer) [inline include/mapbox/geojsonvt/tile.hpp]: This should be rewritten using the visitor pattern.

File: include/mapbox/geojsonvt/clip.hpp (hunk 3 above)
Comment (mourner) [inline include/mapbox/geojsonvt/clip.hpp]: Yep :) Just a way to make the code a bit shorter

File: include/mapbox/geojsonvt/clip.hpp [hunk 9]
```diff
@@ -1,46 +1,221 @@
...
+template <uint8_t I>
+inline void
+clipLine(const vt_line_string& line, const double k1, const double k2, vt_multi_line_string& slices) {
+
+    const double dist = line.dist;
+    const size_t len = line.size();
+
+    vt_line_string slice;
+
+    for (size_t i = 0; i < (len - 1); ++i) {
```
Comment (kkaefer) [inline include/mapbox/geojsonvt/clip.hpp]: potential overflow

File: include/mapbox/geojsonvt/clip.hpp [hunk 10]
```diff
@@ -1,46 +1,221 @@
...
+
+template <uint8_t I>
+inline vt_linear_ring clipRing(const vt_linear_ring& ring, double k1, double k2) {
+
+    const size_t len = ring.size();
+
+    vt_linear_ring slice;
+    slice.area = ring.area;
+
+    for (size_t i = 0; i < (len - 1); ++i) {
```
Comment (kkaefer) [inline include/mapbox/geojsonvt/clip.hpp]: potential overflow

**Comment by jfirebaugh** [comment]
`detail::convert(geometry_collection<double>(), ...)` is going to produce a stack overflow. There isn't an explicit `project` overload for `geometry_collection<double>`, and the reason that it even compiles right now is that the `project` overload for `geometry<double>` gets selected, using the implicit conversion from `geometry_collection<double>` to `geometry<double>`. 😱

**Pull request merged**